![Erudio logo](img/erudio-logo-small.png)

# Using Dates, Times, and Datetimes in Python

This lesson aims at understanding more about the concepts of date and time in Python. In this module, we will explore the intricacies of working with dates and times in Python, covering fundamental concepts such as datetime objects, timedelta, and the datetime module. We will delve into the challenges posed by timezones, discussing how to properly account for them in your applications to ensure accurate and reliable time-based functionality.

Concepts covered:
* datetimes
* dates / times
* timedeltas. 
* timezones

Python provides solid, basic handling of timestamps and other time and calendar functions.  

With a little bit of assistance from third-party libraries, we can manage more intricate behavior of time zones and the complications attached thereto.

In almost any domain of software development, dealing with time-related indicators will be essential, and this course will give you a solid foundation for your work.

Whether you are a beginner looking to understand the basics or an experienced developer seeking to enhance your proficiency, this lesson aims to provide you with a comprehensive understanding of how Python facilitates the manipulation and management of dates, datetimes, and timezones.

# Basic Datetimes

The Python standard library comes with a module `datetime` that handles working with dates and times.  It contains the classes `datetime`, `time`, `date`, and `timedelta` that allow constructing, representing and working with time-oriented objects.  However, issues about timezones are deferred until the next lesson of this course.

One oddity about the `datetime` module is shared by a few other (older) standard library modules like `glob`, `array`, `copy`, and `time` and `pprint`.  In each, the main names inside the module you wish to import has the same name as the module itself.  I often trip over whether I used `import glob` or `from glob import glob` at the top of my code.

In [1]:
# Import cessary modules 
import datetime
from pprint import pprint

## Datetime

A `datetime` object represents a complete timestamp or "moment" in time, optionally including time zone information.  There are a number of different ways of constructing these objects and a number of formats they can be represented as.  The actual class constructor takes a tuple of different components of the instant in time.  For example we might want to represent the current time when the code is called.

In [3]:
# Creating a datetime object
ts0 = datetime.datetime.now()
ts0

datetime.datetime(2024, 2, 8, 19, 38, 38, 42120)

Or we can represent a specific time by providing numeric arguments for each component of the time.

In [4]:
ts1 = datetime.datetime(year=2021, month=1, day=20, 
                        hour=12, minute=0, second=0, 
                        microsecond=123_456)
ts1

datetime.datetime(2021, 1, 20, 12, 0, 0, 123456)

In [5]:
# Parameters may be passed purely positionally
ts2 = datetime.datetime(2021, 1, 20, 12, 0, 0, 345_678)
ts2

datetime.datetime(2021, 1, 20, 12, 0, 0, 345678)

Often it is useful to create datetimes from or export to popular string representations. ISO 8601 format is the best choice, of course, but your requirement might be different.

In [7]:
# Equivalent to: '2021-01-20T12:00:00.678901'
ts3 = datetime.datetime.fromisoformat('2021-01-20 12:00:00.345678')
ts3

datetime.datetime(2021, 1, 20, 12, 0, 0, 345678)

In [8]:
ts0.isoformat()

'2024-02-08T19:38:38.042120'

Custom formats can be defined using the format codes from C-family languages (ANSI X3.159-1989).  For Python, you can read documentation at:

> https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior

There are a large number of these format codes, many of them locale aware.  For example, `%A` will produce one of "Sunday", "Monday", etc. in the locale `en_US`, but will produce "Sonntag", "Montag", etc. for `de_DE`.  The names `strptime()` and `strftime()` are also taken from that old C standard.

An example of output.

In [9]:
import locale
print(locale.getdefaultlocale())

fmt = '%A %b %d; %Y; %I:%M:%S %p'
datetime.datetime.strftime(ts0, fmt)

('en_IN', 'cp1252')


'Thursday Feb 08; 2024; 07:38:38 PM'

Reading in these formats uses the same codes.  The module is only semi-smart about validation.

In [10]:
# Good datetime object
strptime = datetime.datetime.strptime
strptime('Wednesday Aug 12; 2020; 05:29:12 PM', fmt)

datetime.datetime(2020, 8, 12, 17, 29, 12)

In [11]:
# Still "good" even though wrong weekday
strptime('Monday Aug 12; 2020; 05:29:12 PM', fmt)

datetime.datetime(2020, 8, 12, 17, 29, 12)

Some validation is performed though.

In [12]:
# It DOES check that it is a day name at all
try:
    strptime('Humpday Aug 08; 2020; 05:29:12 PM', fmt)
except ValueError as err:
    pprint(str(err), width=54)

("time data 'Humpday Aug 08; 2020; 05:29:12 PM' does "
 "not match format '%A %b %d; %Y; %I:%M:%S %p'")


In [13]:
# Day number not in month
try:
    strptime('Thursday Sep 31; 2020; 05:29:12 PM', fmt)
except ValueError as err:
    print(err)

day is out of range for month


# Basic Datetimes
## Comparing Datetimes

Datetime objects are comparable and hashable.  This makes it easy both to compare different datetimes and to use these objects as elements in sets or keys in dictionaries.  When you try out this notebookBasic DatetimesBasic DatetimesBasic DatetimesBasic DatetimesBasic DatetimesBasic Datetimes## Comparing Datetimes

Inequalities and equality:

In [14]:
print(ts0, ts1, sep='\n')
ts0 < ts1

2024-02-08 19:38:38.042120
2021-01-20 12:00:00.123456


False

In [15]:
print(ts2, ts3, sep='\n')
ts2 == ts3

2021-01-20 12:00:00.345678
2021-01-20 12:00:00.345678


True

In [16]:
dates = [ts0, ts1, ts2, ts3]
objs = {id(ts) for ts in dates}
distinct = set(dates)
print(objs)
distinct

{1789656710336, 1789656696368, 1789656698000, 1789656709568}


{datetime.datetime(2021, 1, 20, 12, 0, 0, 123456),
 datetime.datetime(2021, 1, 20, 12, 0, 0, 345678),
 datetime.datetime(2024, 2, 8, 19, 38, 38, 42120)}

dates = [ts0, ts1, ts2, ts3] objs = {id(ts) for ts in dates} distinct = set(dates) print(objs) distinctSimilarly with dictionary keys.

In [17]:
{ts0: "Hope", ts1: "Change", ts2: "Progress"}

{datetime.datetime(2024, 2, 8, 19, 38, 38, 42120): 'Hope',
 datetime.datetime(2021, 1, 20, 12, 0, 0, 123456): 'Change',
 datetime.datetime(2021, 1, 20, 12, 0, 0, 345678): 'Progress'}

## Date and Time

The separate classes `date` and `time` represent those named components of an overall `datetime`.  However, these time-related datatypes are simply different, and require explicit conversion to or from a `datetime`.  As with full datetime objects, we can construct dates and times in a number of ways.

The most basic constructors just give componets.

In [19]:
print("Date:", datetime.date(1941, 12, 7))
print("Time:", datetime.time(7, 55, 0))

Date: 1941-12-07
Time: 07:55:00


One way of constructing a date (or a datetime) is in terms of "seconds since the beginning of (Unix) time." For a date alone, this only gets you "some time during that day."

In [21]:
import time
now = time.time()
# Seconds since the epoch
print(f" Seconds: {now:,}")
# What timestamp is that?
print("Datetime:", datetime.datetime.fromtimestamp(now))
# What day is that?
print("    Date:", datetime.date.fromtimestamp(now))

 Seconds: 1,707,402,600.2836485
Datetime: 2024-02-08 20:00:00.283648
    Date: 2024-02-08


Inequalities and equality:

In [12]:
print(ts0, ts1, sep='\n')
ts0 < ts1

2020-08-15 16:08:11.183949
2021-01-20 12:00:00.123456


True

In [13]:
print(ts2, ts3, sep='\n')
ts2 == ts3

2021-01-20 12:00:00.345678
2021-01-20 12:00:00.345678


True


Datetime objects are hashable.  Equality, of course, means that only one equal object is in a set.

In [14]:
dates = [ts0, ts1, ts2, ts3]
objs = {id(ts) for ts in dates}
distinct = set(dates)
print(objs)
distinct

{140149750458912, 140149750513408, 140149750512784, 140149750512112}


{datetime.datetime(2020, 8, 15, 16, 8, 11, 183949),
 datetime.datetime(2021, 1, 20, 12, 0, 0, 123456),
 datetime.datetime(2021, 1, 20, 12, 0, 0, 345678)}

Similarly with dictionary keys.

In [15]:
{ts0: "Hope", ts1: "Change", ts2: "Progress"}

{datetime.datetime(2020, 8, 15, 16, 8, 11, 183949): 'Hope',
 datetime.datetime(2021, 1, 20, 12, 0, 0, 123456): 'Change',
 datetime.datetime(2021, 1, 20, 12, 0, 0, 345678): 'Progress'}

If we pick an arbitrary number, lets say "0" as the seconds-since-epoch, we will get a date (or datetime; but **not** a time for this method).

In [22]:
datetime.date.fromtimestamp(0)

datetime.date(1970, 1, 1)

This is a somewhat odd date, given that Unix time started at the stroke of midnight on January 1, 1970.  The key thing is Unix time start in UTC timezone, and my local time writing this is EST.

Negative numbers are perfectly fine, But does it work with the fromtimestamp function? Let's see

In [26]:
long_ago = datetime.date.fromtimestamp(-6_857_300_000)
long_ago

OSError: [Errno 22] Invalid argument

The issue we're encountering is related to the limitations of the **fromtimestamp** method in the datetime module when dealing with negative timestamps.

The fromtimestamp method expects a timestamp representing the number of seconds since the epoch (January 1, 1970, 00:00:00 UTC). However, when you provide a negative timestamp, it attempts to interpret it as a date before the epoch, which is not supported.

If you want to work with a date that is far in the past or future, you can use the datetime constructor directly, specifying the year, month, and day. Here's an example for a date far in the past:

In [28]:
from datetime import datetime, timedelta

long_ago = datetime(1970, 1, 1) + timedelta(seconds=-6_857_300_000)
print(long_ago)


1752-09-13 02:26:40


This gives us another fun date, in that in England, and all its colonies of the time, the calendar was adjusted in 1752 between the Julian and Gregorian calendars, meaning that the dates between September 2 and September 14 did not occur that year. Most of the rest of Europe had made the switch back in 1582, but there were some quibbles, and large wars, going on in that part of the world, known as the "Reformation" and "Counter-Reformation."

Calendrics are a strange and confusing story. We might jump forward from that date that perhaps did not exist to my our own birth date.


In [29]:
one_day = long_ago.replace(year=1964)
one_day

datetime.datetime(1964, 9, 13, 2, 26, 40)

## Converting Among Time Objects

We need to use explicit methods to move among the different kinds of date/time objects.  A datetime is the more universal object; I rarely bother with a date or time alone except perhaps as a final conversion step.

In [30]:
print(" Now:", ts0)
print("Date:", ts0.date())
print("Time:", ts0.time())

 Now: 2024-02-08 19:38:38.042120
Date: 2024-02-08
Time: 19:38:38.042120


From a datetime it is perfectly sensible to extract a date or a time, but the other direction is either lossy or impossible.  In concept, you could destructure component to generate a datetime.

In [35]:
import datetime
y, mon, d = long_ago.year, long_ago.month, long_ago.day
h, m, s = (12, 1, 30)
datetime.datetime(2021, 1, 20, 12, 0, 0, 345_678)
datetime.datetime(y, mon, d, h, m, s).isoformat()

'1752-09-13T12:01:30'

In practice, we have a method to put these together.

In [36]:
combine = datetime.datetime.combine
combine(long_ago, datetime.time(12, 1, 30))

datetime.datetime(1752, 9, 13, 12, 1, 30)

## Timedelta

One thing that is often important is measuring the distance between datetimes.  To work with durations—or "deltas"—we use the `timedelta` class.  This is also produced as the result of subtraction between objects.

In [38]:
thousand_days = datetime.timedelta(days=1000)
print(thousand_days)

1000 days, 0:00:00


In [39]:
five_minutes = datetime.timedelta(minutes=5)
print(five_minutes)

0:05:00


Timedeltas can be added of subtracted from each other, with reasonable behavior.

In [40]:
after_a_while = thousand_days + five_minutes
print(after_a_while)

1000 days, 0:05:00


In [41]:
a_bit_less = thousand_days - five_minutes
a_bit_less

datetime.timedelta(days=999, seconds=86100)

In [42]:
print(4 * after_a_while)

4000 days, 0:20:00


### Equalities and inequalities behave as you would expect.

In [43]:
after_a_while == datetime.timedelta(days=1000, minutes=5)

True

In [44]:
a_bit_less < after_a_while

True

There are a number of units available in constructing timedeltas, but the most coarse is days, not months, years, or centuries.  Units are not limited to the number of their cyclicities (24 hours, 60 seconds, etc).

In [45]:
hours_50k = datetime.timedelta(hours=50_000)
print(hours_50k)

2083 days, 8:00:00


In [46]:
hours_50k <= thousand_days + thousand_days

False

### Addition and Subtraction

Subtracting one datetime, or date, from another produces a timedelta.  Long increments will account for leap days where relevant, even though the units will not include years (just more days).  For days, the delta will be whole days only, and other units will not appear.

In [47]:
one_day - long_ago

datetime.timedelta(days=77431)

In [48]:
ts0 - ts1

datetime.timedelta(days=1114, seconds=27517, microseconds=918664)

Using times is more limited, correctly so. Subtracting times seems like it makes sense, but of course they roll over at midnight each night, which causes problems.

In [53]:
t1 = datetime.time(hour=18, minute=25, second=35)
t2 = datetime.time(hour=20, minute=0, second=0)
try:
    t2 - t1
except TypeError as err:
    print(err)

unsupported operand type(s) for -: 'datetime.time' and 'datetime.time'


The solution is to use datetimes rather than only times.

In [54]:
today = datetime.date.today()
ts4 = combine(today, t1)
ts5 = combine(today, t2)
print("Waiting:", ts5 - ts4)

Waiting: 1:34:25


Code like that just above could go wrong if times you want the difference between cross midnight.  This is something that can commonly occur, for example, when writing to log files. However, the solution is simple.  Do not fix "today" as I did above, but rather call `today()` at the moment the entry is created.  

In other cases, it will not be "today" you want; but just think about appropriate date components.

In [55]:
tomorrow = today + datetime.timedelta(days=1)
late_night = datetime.time(23, 59, 30)
early_next = datetime.time(0, 2, 15)
print(late_night)
print(early_next)

23:59:30
00:02:15


In [56]:
ts6 = combine(today, late_night)
ts7 = combine(tomorrow, early_next)
ts7 - ts6

datetime.timedelta(seconds=165)

Often the use of a timedelta is not simply describing a duration, but also applying it to an existing datetime or date. We cannot similarly apply it to a time because of the wrap-around issues (even though some such operations would not wrap, it is excluded at the level of the datatype).

In [57]:
print(ts1)
print(after_a_while)
ts1 + after_a_while

2021-01-20 12:00:00.123456
1000 days, 0:05:00


datetime.datetime(2023, 10, 17, 12, 5, 0, 123456)

In [58]:
print(today)
print(today + after_a_while)

2024-02-08
2026-11-04


Some deltas will not change the date.

In [59]:
print(today + datetime.timedelta(seconds=1))

2024-02-08


But not for a time. Not even one that is not near midnight.

In [60]:
print(t1)
try: 
    t1 + datetime.timedelta(seconds=1)
except Exception as err:
    print(err)

18:25:35
unsupported operand type(s) for +: 'datetime.time' and 'datetime.timedelta'


#### Why the above code erroed out?
We are attempting to add a timedelta to a variable t1, and t1 is a datetime object. However, there is a minor syntax error in your code. The datetime module needs to be imported to use timedelta. Here's the corrected code:

In [61]:
from datetime import datetime, timedelta

t1 = datetime.now()
print(t1)

try:
    t1 + timedelta(seconds=1)
except Exception as err:
    print(err)


2024-02-08 20:46:10.316040


## More on this as we explore the different concepts

# Exercise 

For this exercise, you will need to consult a reference on date formatting codes, such as:

> https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior

Since the conventions are shared by a great many programming languages and tools, other references are equally accurate.

Your goal is to write a function, `future()` that will take a string that describes a timestamp, calculates the timestamp at a certain distance into the future, and prints out the future date in the same format.  The signature of your function will be:

```python
def future(ts: str, days: int=0, hours: int=0, minutes: int=0) -> str:
    ...
```

In a basic version of this solution, simply match the format in the examples shown.  If you would like to practice a little bit more, add an optional argument `fmt` that allows accepting other timestamp formats.

```python
>>> future("9 April 2019; 07:08:09", days=100, minutes=100)
'18 July 2019; 08:48:09'

>>> future("10 May 2020; 23:24:25", hours=10_000)
'01 July 2021; 15:24:25'

>>> future("11 June 2021; 11:12:13", hours=10**4, minutes=-10**6)
'06 September 2020; 16:32:13'
```

# Setup

In [63]:
dt = datetime.now()
date = dt.date()
time = dt.time()
delta = timedelta(hours=2)

In [64]:
from datetime import datetime, timedelta

# This function has correct signature, but wrong behavior
def future(ts, days=0, hours=0, minutes=0):
    return datetime.now().isoformat()

# Solution

In [65]:
def future(ts, days=0, hours=0, minutes=0,
           fmt = '%d %B %Y; %H:%M:%S'):
    dt = datetime.strptime(ts, fmt)
    delta = timedelta(days=days, hours=hours, minutes=minutes)
    return datetime.strftime(dt + delta, fmt)

In [66]:
def test_conversion():
    ts = "9 April 2019; 07:08:09"
    new = future(ts, days=100, minutes=100)
    assert new == '18 July 2019; 08:48:09', new
    ts = "10 May 2020; 23:24:25"
    new = future(ts, hours=10_000)
    assert new == '01 July 2021; 15:24:25', new
    ts = "11 June 2021; 11:12:13"
    new = future(ts, hours=10**4, minutes=-10**6)
    assert new == '06 September 2020; 16:32:13', new

In [67]:
test_conversion()

-------------
Materials licensed under [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/) by the authors